### PR TITLE
feat: typed errors for templates use cases

### PR DESCRIPTION
**Pull request by lkingland**
Typed errors when dealing with both embedded and external template repositories.

### CODE REVIEW

File: client_test.go [hunk 1]
```diff
@@ -180,37 +180,107 @@ func TestExtensibleTemplates(t *testing.T) {
 
 	// Create a new client with a path to the extensible templates
 	client := bosonFunc.New(
-		bosonFunc.WithTemplates("testdata/templates"),
+		bosonFunc.WithTemplates("testdata/repositories"),
```
Comment (lkingland) [inline client_test.go:183]: This will be renamed to "packages" now that we have decided on a rename of --tempates per #381

**Review by lance — COMMENTED** [review]
I like the typed errors. That's a nice addition. I want to make sure we're in agreement on a repository/package namespacing.

File: client_test.go (hunk 1 above)
Comment (lkingland) [reply client_test.go:183]: Opeining this as a separate issue, as it has a fair bit of discussion around it.

**Comment by lkingland** [comment]
> I like the typed errors. That's a nice addition. I want to make sure we're in agreement on a repository/package namespacing.

Sounds good.  Our agreed-upon solution is in #384

**Review by lance — APPROVED** [review]

**Review by zroubalik — APPROVED** [review]

**Pull request merged**